### PR TITLE
Storage subscription

### PR DESCRIPTION
**Pull request by cube2222**

### CODE REVIEW

File: streaming/storage/storage.go [hunk 1]
```diff
@@ -28,7 +27,7 @@ func NewBadgerStorage(db *badger.DB) *BadgerStorage {
 }
 
 func (bs *BadgerStorage) BeginTransaction() StateTransaction {
-	//bs.db.DropPrefix()
+	// bs.db.DropPrefix()
```
Comment (JasiekChomiak) [inline streaming/storage/storage.go]: If you are modifying the comment you might as well remove this line, I don't really think it's necessary to keep it there

File: streaming/storage/storage.go [hunk 2]
```diff
@@ -45,15 +44,15 @@ func (bs *BadgerStorage) WithPrefix(prefix []byte) Storage {
...
-
-	err := bs.db.Subscribe(ctx, cb, bs.prefix)
-	if err != nil {
-		return errors.Wrap(err, "couldn't subscribe to db")
-	}
-
-	return nil
+func (bs *BadgerStorage) Subscribe(ctx context.Context) *Subscription {
+	return NewSubscription(ctx, func(ctx context.Context, changes chan<- struct{}) error {
+		return bs.db.Subscribe(ctx, func(kv *badger.KVList) error {
```
Comment (JasiekChomiak) [inline streaming/storage/storage.go]: Use KeyValueList defined above instead of *badger.KVList

File: streaming/storage/subscription.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+	errs := make(chan error, 1) // Buffered so there won't be a goroutine leak if nobody reads the error.
+	sub := &Subscription{
+		cancel:  cancel,
+		changes: changes,
+		errors:  errs,
+		wg:      sync.WaitGroup{},
+	}
+
+	sub.wg.Add(1)
+	go func() {
```
Comment (JasiekChomiak) [inline streaming/storage/subscription.go]: I think that a constructor that starts a goroutine is quite peculiar. Maybe it's a Go construct that I haven't seen. You could add a short comment explaining why you do it in this specific place.

File: streaming/storage/storage.go (hunk 1 above)
Comment (cube2222) [reply streaming/storage/storage.go]: Autoformat, but sure

File: streaming/storage/storage.go (hunk 2 above)
Comment (cube2222) [reply streaming/storage/storage.go]: removed the type instead

File: streaming/storage/subscription.go (hunk 3 above)
Comment (cube2222) [reply streaming/storage/subscription.go]: Added comment

**Pull request merged**